### PR TITLE
PP-8802 Add Org URL - Add tests, validators and call to Stripe and call to Admin Users

### PR DESCRIPTION
**Pull request by iqbalgds**
- Refactor the `Add Org URL` controller
   - use the existing pattern of a seperate `get` and `post` controllers.
   - Move the files into its own `kyc/organisation-url` sub-folder
- Add checks to see if the user is doing the following - switching PSP
  or  adding KYC data
  - Use these check to decide which page to redirect to depending
    if the user has done a successful or error journey.
- Add unit tests for the `get` and `post` controller.
- Update Validation to use the URL validator
- Add call to update Stripe with the Organisation URL
  - Created a new StripeAccount object to do this.
- Add call to admin users service to save organisation URL.

### CODE REVIEW

**Review by kbottla — CHANGES_REQUESTED** [review]

File: app/controllers/kyc/get.controller.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  const isSwitchingCredentials = isSwitchingCredentialsRoute(req)
+  const collectingAdditionalKycData = isAdditionalKycDataRoute(req)
+  const currentCredential = getCurrentCredential(req.account)
+
+  const stripeAccountSetup = req.account.connectorGatewayAccountStripeProgress
+
+  if (!stripeAccountSetup) {
+    return next(new Error('Stripe setup progress is not available on request'))
+  }
+  if (stripeAccountSetup.organisationUrl) {
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/get.controller.js:17]: Done - thanks

File: app/controllers/kyc/get.controller.test.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    }
+    res = {
+      redirect: sinon.spy(),
+      render: sinon.spy()
+    }
+    next = sinon.spy()
+  })
+
+  it('should render error if organisation URL are already provided', async () => {
+    req.account.connectorGatewayAccountStripeProgress = { organisationUrl: true }
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/get.controller.test.js:28]: Done - thanks

File: app/controllers/kyc/get.controller.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+const { response } = require('../../utils/response')
+const { isSwitchingCredentialsRoute, isAdditionalKycDataRoute, getCurrentCredential } = require('../../utils/credentials')
+const { getAlreadySubmittedErrorPageData } = require('../stripe-setup/stripe-setup.util')
+
+module.exports = (req, res, next) => {
+  const isSwitchingCredentials = isSwitchingCredentialsRoute(req)
+  const collectingAdditionalKycData = isAdditionalKycDataRoute(req)
+  const currentCredential = getCurrentCredential(req.account)
+
+  const stripeAccountSetup = req.account.connectorGatewayAccountStripeProgress
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/get.controller.js:12]: Done

File: app/routes.js [hunk 4]
```diff
@@ -429,7 +429,8 @@ module.exports.bind = function (app) {
   account.post(paymentLinks.manage.editMetadata, permission('tokens:create'), paymentLinksController.postUpdateReportingColumn.editMetadata)
   account.post(paymentLinks.manage.deleteMetadata, permission('tokens:create'), paymentLinksController.postUpdateReportingColumn.deleteMetadata)
 
-  account.get(kyc.organisationUrl, permission('merchant-details:update'), kycController.organisationUrl.getUrl)
+  account.get(kyc.organisationUrl, permission('merchant-details:update'), kycController.get)
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/routes.js:432]: - Added middleware to `restrictToStripeAccountContext `
- Moved the controller and tests to a sub-folder: `kyc/organisation-url`

File: app/controllers/kyc/post.controller.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      })
+
+      if (isSwitchingCredentials) {
+        return res.redirect(303, formatAccountPathsFor(paths.account.switchPSP.index, req.account.external_id))
+      } else if (collectingAdditionalKycData) {
+        req.flash('generic', 'Organisation website address added successfully')
+        return res.redirect(303, formatAccountPathsFor(paths.account.yourPsp.index, req.account && req.account.external_id, currentCredential.external_id))
+      }
+
+      return res.redirect(303, formatAccountPathsFor(paths.account.stripe.addPspAccountDetails, req.account && req.account.external_id))
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/post.controller.js:64]: Done

File: app/controllers/kyc/post.controller.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    pageData['errors'] = errors
+
+    return response(req, res, 'kyc/organisation-url', {
+      ...pageData, isSwitchingCredentials, collectingAdditionalKycData, currentCredential
+    })
+  } else {
+    try {
+      const stripeAccountId = await getStripeAccountId(req.account, isSwitchingCredentials, req.correlationId)
+
+      await updateCompany(stripeAccountId, { business_profile: { url: organisationUrl } })
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/post.controller.js]: As discussed, created a a new `StripeAcccount` class to update the `business_profile > url`

File: app/controllers/kyc/post.controller.test.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        stripeAccount: {
+          stripeAccountId: 'acct_123example123'
+        }
+      }
+    }
+    next = sinon.spy()
+    updateCompanyMock = sinon.spy(() => Promise.resolve())
+  })
+
+  it('update Stripe company URL, update admin users client and then redirect to add PSP account details', async () => {
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/post.controller.test.js]: Will create as a separate PR.

File: app/controllers/kyc/post.controller.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
```
[the remark this answers is not in the document]
Comment (iqbalgds) [reply app/controllers/kyc/post.controller.js]: This has been added - thanks

File: app/controllers/kyc/organisation-url/get.controller.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+const { response } = require('../../../utils/response')
+const { isSwitchingCredentialsRoute, isAdditionalKycDataRoute, getCurrentCredential } = require('../../../utils/credentials')
+
+module.exports = (req, res, next) => {
+  const isSwitchingCredentials = isSwitchingCredentialsRoute(req)
+  const collectingAdditionalKycData = isAdditionalKycDataRoute(req)
+  const currentCredential = getCurrentCredential(req.account)
+
+  if (collectingAdditionalKycData && !req.account.requires_additional_kyc_data) {
+    return next(new Error('requires_additional_kyc_data flag is false on request'))
```
Comment (kbottla) [inline app/controllers/kyc/organisation-url/get.controller.js:12]:
```suggestion
    return next(new Error('requires_additional_kyc_data flag is not enabled for gateway account'))
```

Comment (iqbalgds) [reply app/controllers/kyc/organisation-url/get.controller.js]: Thanks - have updated the file and test.

**Review by kbottla — APPROVED** [review]

**Pull request merged**